exporter/exporterparser/zipkin.go

```diff
@@ -106,89 +110,250 @@ func ZipkinExportersFromYAML(config []byte) (tes []exporter.TraceExporter, doneF
+// are per-span-Node attributes instead of global/system variables.
+// The alternative is to create a single exporter for every single combination
+// but this wastes resources i.e. an HTTP client for every single combination
+// but also requires the exporter to be changed entirely as per
+// https://github.com/census-instrumentation/opencensus-go/issues/959
+//
+// TODO: (@odeke-em) whenever we come to consensus with the OpenCensus-Go repository
+// on the Zipkin exporter and they have the same logic, then delete all the code
+// below here to allow per-span configuration changes.
 
-	// Try looking up if we already created the exporter.
-	if exp, ok := ze.serviceNameToExporter[key]; ok && exp != nil {
-		return exp
+const (
+	statusCodeTagKey        = "error"
+	statusDescriptionTagKey = "opencensus.status_description"
+)
+
+var (
+	sampledTrue    = true
+	canonicalCodes = [...]string{
+		"OK",
+		"CANCELLED",
+		"UNKNOWN",
+		"INVALID_ARGUMENT",
+		"DEADLINE_EXCEEDED",
+		"NOT_FOUND",
+		"ALREADY_EXISTS",
+		"PERMISSION_DENIED",
+		"RESOURCE_EXHAUSTED",
+		"FAILED_PRECONDITION",
+		"ABORTED",
+		"OUT_OF_RANGE",
+		"UNIMPLEMENTED",
+		"INTERNAL",
+		"UNAVAILABLE",
+		"DATA_LOSS",
+		"UNAUTHENTICATED",
 	}
+)
 
-	localEndpointURI := ze.defaultLocalEndpointURI
-	if ipv4 := node.Attributes["ipv4"]; ipv4 != "" {
-		localEndpointURI = ipv4
-	} else if ipv6 := node.Attributes["ipv6"]; ipv6 != "" {
-		localEndpointURI = ipv6
+func canonicalCodeString(code int32) string {
+	if code < 0 || int(code) >= len(canonicalCodes) {
+		return "error code " + strconv.FormatInt(int64(code), 10)
 	}
+	return canonicalCodes[code]
+}
 
-	// Otherwise freshly create that Zipkin Exporter.
-	localEndpoint, err := openzipkin.NewEndpoint(serviceName, localEndpointURI)
-	if err != nil {
-		return ze.defaultExporter
+func convertTraceID(t trace.TraceID) zipkinmodel.TraceID {
+	return zipkinmodel.TraceID{
+		High: binary.BigEndian.Uint64(t[:8]),
+		Low:  binary.BigEndian.Uint64(t[8:]),
 	}
+}
 
-	reporter := http.NewReporter(ze.endpointURI)
-	exporter := zipkin.NewExporter(reporter, localEndpoint)
+func convertSpanID(s trace.SpanID) zipkinmodel.ID {
+	return zipkinmodel.ID(binary.BigEndian.Uint64(s[:]))
+}
 
-	// Now memoize the created exporter for later use.
-	ze.serviceNameToExporter[key] = exporter
-	ze.shutdownFns = append(ze.shutdownFns, reporter.Close)
+func spanKind(s *trace.SpanData) zipkinmodel.Kind {
+	switch s.SpanKind {
+	case trace.SpanKindClient:
+		return zipkinmodel.Client
+	case trace.SpanKindServer:
+		return zipkinmodel.Server
+	}
+	return zipkinmodel.Undetermined
+}
 
-	return exporter
+func remoteZipkinServiceNamePrefix(fieldName string) string {
+	return "zipkin.remoteEndpoint." + fieldName
 }
 
-func (ze *zipkinExporter) stop() error {
-	ze.mu.Lock()
-	defer ze.mu.Unlock()
+func (ze *zipkinExporter) serviceNameOrDefault(node *commonpb.Node) string {
+	ze.mu.RLock()
+	defaultServiceName := ze.defaultServiceName
```

Comment:
Per conversation on previous PR the mutex is used here to protect the method during a call to stop(), however, the code as it is will give the impression that `ze.defaultServiceName` can actually change which is not the case (did I miss it somewhere?). If this is to prevent execution on reporter on the case of a call to stop() then using a RW lock with a int flag to signal when it is being closed (or already closed) seems reasonable.

Reply:
@pjanotti thanks for raising that, it was more of a defensibility for safety which as I mentioned let's try to ensure that for accesses we first protect them then perhaps later on remove locks. I've removed it and updated the code, PTAL and thanks again.